packages/vhd-lib/src/createReadableSparseStream.js

```diff
@@ -17,38 +18,65 @@ import { set as setBitmap } from './_bitmap'
 const VHD_BLOCK_SIZE_SECTORS = VHD_BLOCK_SIZE_BYTES / SECTOR_SIZE
 
 /**
+ * Looks once backwards to collect the last fragment of each VHD block (they could be interleaved),
+ * then allocates the blocks in a forwards pass.
  * @returns currentVhdPositionSector the first free sector after the data
  */
 function createBAT(
   firstBlockPosition,
-  blockAddressList,
+  fragmentLogicAddressList,
   ratio,
   bat,
   bitmapSize
 ) {
   let currentVhdPositionSector = firstBlockPosition / SECTOR_SIZE
-  blockAddressList.forEach(blockPosition => {
-    assert.strictEqual(blockPosition % SECTOR_SIZE, 0)
-    const vhdTableIndex = Math.floor(blockPosition / VHD_BLOCK_SIZE_BYTES)
-    if (bat.readUInt32BE(vhdTableIndex * 4) === BLOCK_UNUSED) {
-      bat.writeUInt32BE(currentVhdPositionSector, vhdTableIndex * 4)
-      currentVhdPositionSector +=
-        (bitmapSize + VHD_BLOCK_SIZE_BYTES) / SECTOR_SIZE
+  const lastFragmentPerBlock = new Map()
+  forEachRight(fragmentLogicAddressList, fragmentLogicAddress => {
+    assert.strictEqual(fragmentLogicAddress % SECTOR_SIZE, 0)
+    const vhdTableIndex = Math.floor(
+      fragmentLogicAddress / VHD_BLOCK_SIZE_BYTES
+    )
+    if (!lastFragmentPerBlock.has(vhdTableIndex)) {
+      lastFragmentPerBlock.set(vhdTableIndex, fragmentLogicAddress)
     }
   })
```

Comment:
Wouldn't it be easier to iterate forward and to always assign the new address?

Reply:
This may also remove the need of `forEachRight` for the next iteration.

Reply:
I don't think so, because the VMDK grain representing a VHD block might be interleaved.
let's say the VHD blocks `A` and `B` are fragmented in 3 VMDK grains each and that the VMDK stream present the grains in this order (letters are VHD, numbers are VMDK):
`A3 B1 A1 B2 A2`

The VHD BAT using the least intermediate memory possible is `B A` because I can only write the block A on the stream when I receive `A1`, and it occurs after the last Block of `B`. Note that B3 doesn't even appears in the VMDK, its bits in the VHD block bitmap will just stay at zero, and the VHD block will be dumped on the VHD stream when encountering `B2` because we know it's the last time we see a block from `B`.